src/System.Private.Xml/src/System/Xml/XPath/XPathNodeIterator.cs

```diff
@@ -32,17 +33,23 @@ public virtual int Count
                 return count;
             }
         }
+
         public virtual IEnumerator GetEnumerator()
         {
             return new Enumerator(this);
         }
 
+        IEnumerator<XPathNavigator> IEnumerable<XPathNavigator>.GetEnumerator()
+        {
+            return new Enumerator(this);
+        }
```

Comment:
If I create a class that inherits from `XPathNodeIterator` and overrides `GetEnumerator()` to return a custom enumerator, that custom enumerator won't be used here. Is that the expected behavior?

Reply:
@svick I think this will be consistent with what we do in `List<T>` - for code which is already compiled this will be no issue as it will still call to your override - for new code written on top of library which overrides GetEnumerator this might be an inconvenience but I'm not sure if extra logic here to check for the type is worth those few people who override enumerator - especially they can add a new override to fix this. Do you have any specific cases where this might be breaking?

IMO this is ok but I do not feel strongly about it. thoughts?

Reply:
> for code which is already compiled this will be no issue as it will still call to your override

Not sure if it makes a difference, but there's actually one case where this isn't true: if the already compiled code checks for `IEnumerable<XPathNavigator>`, which is exactly what `Enumerable.Cast<T>()` does:

https://github.com/dotnet/corefx/blob/0101324f1ac8555c910bbbfb46c511bb86f0fbe4/src/System.Linq/src/System/Linq/Cast.cs#L33-L47

So, code like the following could change behavior without recompilation:

```c#
foreach (var nav in myIterator.Cast<XPathNavigator>())
{
    // use nav here
}
```

Consider that using `Cast<T>` like this could be fairly likely, exactly because `XPathNodeIterator` didn't implement `IEnumerable<XPathNavigator>` until now.

> for new code written on top of library which overrides GetEnumerator this might be an inconvenience but I'm not sure if extra logic here to check for the type is worth those few people who override enumerator - especially they can add a new override to fix this.

Sounds reasonable to me.

> Do you have any specific cases where this might be breaking?

I do not.

Reply:
I'd look for existing examples from 1.x -> 2.0 to see what we did for similar situations.

Reply:
@justinvp You remember any? You did GroupCollection - but it wasn't virtual. Can you think of any other examples?

Reply:
I don't recall any examples with a virtual GetEnumerator.

Reply:
Note that there's at least one built-in subclass of `XPathNodeIterator` ([`XPathArrayIterator`](https://github.com/dotnet/corefx/blob/master/src/System.Private.Xml/src/System/Xml/XPath/Internal/XPathArrayIterator.cs)) that overrides `GetEnumerator`:

https://github.com/dotnet/corefx/blob/a011ef1f49a4409c5ed2cb04b9aa7a89476127df/src/System.Private.Xml/src/System/Xml/XPath/Internal/XPathArrayIterator.cs#L83-L86

I think you'll want to implement the explicitly implemented generic `GetEnumerator` on `XPathNodeIterator` in terms of the non-generic virtual `GetEnumerator`, e.g.:

```c#
public abstract class XPathNodeIterator : ICloneable, IEnumerable<XPathNavigator>
{
    ...

    public virtual IEnumerator GetEnumerator()
    {
        return new Enumerator(this);
    }

    IEnumerator<XPathNavigator> IEnumerable<XPathNavigator>.GetEnumerator()
    {
        IEnumerator enumerator = GetEnumerator();
        return enumerator is IEnumerator<XPathNavigator> e ? e : Cast(enumerator);
    }

    private static IEnumerator<XPathNavigator> Cast(IEnumerator source)
    {
        while (source.MoveNext())
        {
            yield return (XPathNavigator)source.Current;
        }
    }

    ...
}
```

This would allow subclasses that override `GetEnumerator`, such as `XPathArrayIterator`, to continue to work.